✨ `Bing` for the search engine

jfvillablanca opened the pull request:
## What does this PR do?

<!-- MANDATORY -->
Adds `Bing` search engine to websurfx
<!-- explain the changes in your PR, algorithms, design, architecture -->

## Why is this change important?
Expand the search results horizon with `Bing` which helps improve user experience.
<!-- MANDATORY -->

<!-- explain the motivation behind your PR -->

## How to test this PR locally?

<!-- commands to run the tests or instructions to test the changes-->
Enable `Bing` in config.lua or via the settings UI.
## Author's checklist

<!-- additional notes for reviewers -->

## Related issues
Closes #316 
<!--
Closes #234
-->

neon-mmd commented on src/engines/bing.rs:
```diff
@@ -0,0 +1,109 @@
+            }
+        };
+
+        let header_map = HeaderMap::try_from(&HashMap::from([
+            ("USER_AGENT".to_string(), user_agent.to_string()),
+            ("REFERER".to_string(), "https://google.com/".to_string()),
+            (
+                "CONTENT_TYPE".to_string(),
+                "application/x-www-form-urlencoded".to_string(),
+            ),
+        ]))
+        .change_context(EngineError::UnexpectedError)?;
```
I think I would suggest adding a `cookie` header because this helps prevent being detected as a bot by the upstream search engine. 

Also, the cookie for bing will look something like the following:
```
MUID=2D0BD206B856600B0802C1DBB95F61B9; _EDGE_V=1; MUIDB=2D0BD206B856600B0802C1DBB95F61B9; SRCHD=AF=NOFORM; SRCHUID=V=2&GUID=B68AA9D86D6F45C5A891D99D8902C941&dmnchg=1; _HPVN=CS=eyJQbiI6eyJDbiI6MiwiU3QiOjAsIlFzIjowLCJQcm9kIjoiUCJ9LCJTYyI6eyJDbiI6MiwiU3QiOjAsIlFzIjowLCJQcm9kIjoiSCJ9LCJReiI6eyJDbiI6MiwiU3QiOjAsIlFzIjowLCJQcm9kIjoiVCJ9LCJBcCI6dHJ1ZSwiTXV0ZSI6dHJ1ZSwiTGFkIjoiMjAyMy0xMi0zMVQwMDowMDowMFoiLCJJb3RkIjowLCJHd2IiOjAsIlRucyI6MCwiRGZ0IjpudWxsLCJNdnMiOjAsIkZsdCI6MCwiSW1wIjo0LCJUb2JuIjowfQ==; _Rwho=u=d; ai_session=TB676+0gO13S3BW5AYVW3d|1704013243633|1704013243633; BCP=AD=0&AL=0&SM=0; ipv6=hit=1704016843140&t=4; _EDGE_S=SID=13B4DE4C147D60EB12DECDB415A661CB&mkt=en-kw; USRLOC=HS=1&ELOC=LAT=29.333396911621094|LON=47.99052810668945|N=Mirqab%2C%20Al%20Asimah|ELT=1|; SRCHUSR=DOB=20231231&T=1704013237000&TPC=1704013248000; bngps=s=0; _UR=QS=0&TQS=0; _RwBf=r=0&ilt=10&ihpd=1&ispd=6&rc=18&rb=0&gb=0&rg=200&pc=15&mtu=0&rbb=0&g=0&cid=&clo=0&v=10&l=2023-12-31T08:00:00.0000000Z&lft=0001-01-01T00:00:00.0000000&aof=0&ard=0001-01-01T00:00:00.0000000&o=2&p=&c=&t=0&s=0001-01-01T00:00:00.0000000+00:00&ts=2023-12-31T09:04:39.0642186+00:00&rwred=0&wls=&wlb=&wle=&ccp=&lka=0&lkt=0&aad=0&TH=; _SS=SID=13B4DE4C147D60EB12DECDB415A661CB&R=18&RB=0&GB=0&RG=200&RP=15; SRCHHPGUSR=SRCHLANG=en&IG=5AFEB9E3084249A9AF2DADF971C5965F&PV=6.6.8&BRW=NOTP&BRH=M&CW=819&CH=703&SCW=1164&SCH=3480&DPR=1.0&UTC=180&DM=1&HV=1704013480&WTS=63839610037&PRVCW=1325&PRVCH=703&NEWWND=0&NRSLT=10c&LSL=0&AS=1&NNT=0&HAP=2&VSRO=0&CHTRSP=3&WEBTHEME=1
```

jfvillablanca replied:
I did not include the cookies because their cookies are very long and generated per-user. How about we include the shorter ones like `_EDGE_v=1`, `SRCHD=AF=NOFORM`, etc, what do you think?

neon-mmd replied:
> I did not include the cookies because their cookies are very long and generated per-user. How about we include the shorter ones like `_EDGE_v=1`, `SRCHD=AF=NOFORM`, etc, what do you think?

Ok I see, I think then I would suggest adding them and checking if everything works still. Also, if there are more cookie parameters then I would suggest exploring them and adding them too. :slightly_smiling_face:

neon-mmd commented on src/engines/bing.rs:
```diff
@@ -0,0 +1,109 @@
+            }
+        }
+
+        let re_span = Regex::new(r#"<span.*?>.*?(?:</span>&nbsp;·|</span>)"#).unwrap();
+        let re_strong = Regex::new(r#"(<strong>|</strong>)"#).unwrap();
+
+        // scrape all the results from the html
+        self.parser
+            .parse_for_results(&document, |title, url, desc| {
+                Some(SearchResult::new(
+                    &re_strong.replace_all(title.inner_html().trim(), ""),
+                    url.value().attr("href").unwrap(),
+                    &re_strong.replace_all(&re_span.replace_all(desc.inner_html().trim(), ""), ""),
+                    &["bing"],
+                ))
+            })
+    }
+}
```
```suggestion

        // scrape all the results from the html
        self.parser
            .parse_for_results(&document, |title, url, desc| {
                Some(SearchResult::new(
                    &re_strong.replace_all(title.inner_html().trim(), ""),
                    url.value().attr("href").unwrap(),
                    desc.inner_html().trim(),
                    &["bing"],
                ))
            })
    }
}
```

I think you don't need to replace `span` and `strong` tags because they will be rendered on the search engine page. Also, having this would also be necessary as this will highlight what matched in the description of the site because of which these results were shown. This can add a little aesthetic style which can provide a good user experience to the user. So I would suggest leaving it :slightly_smiling_face: .

jfvillablanca replied:
For the `strong` tags, I figured it was necessary to sanitize and to maintain uniformity among the search engines, i.e. other search engines do not insert semantic tags on their results. But if we want that in our results, I'll remove it

For the `span` tags, I believe that one is necessary to be removed because it removes a Bing-specific element not relevant to the results

### Screenshot from Bing. The _Web_ text is inside the span which I removed.

![2024-01-01_08-02](https://github.com/neon-mmd/websurfx/assets/31008330/ad4b1b36-6282-4ca8-b648-13e151cc462a)

### Screenshot from Websurfx with the `strong` and `span` tags included
 
![2024-01-01_08-03](https://github.com/neon-mmd/websurfx/assets/31008330/ae2bb279-804f-4085-9ada-4970c217d973)

neon-mmd replied:
> For the `strong` tags, I figured it was necessary to sanitize and to maintain uniformity among the search engines, i.e. other search engines do not insert semantic tags on their results. But if we want that in our results, I'll remove it
> 
> For the `span` tags, I believe that one is necessary to be removed because it removes a Bing-specific element not relevant to the results
> 
> ### Screenshot from Bing. The _Web_ text is inside the span which I removed.
> ![2024-01-01_08-02](https://private-user-images.githubusercontent.com/31008330/293525761-ad4b1b36-6282-4ca8-b648-13e151cc462a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQwODI0MTYsIm5iZiI6MTcwNDA4MjExNiwicGF0aCI6Ii8zMTAwODMzMC8yOTM1MjU3NjEtYWQ0YjFiMzYtNjI4Mi00Y2E4LWI2NDgtMTNlMTUxY2M0NjJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAxVDA0MDgzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3YTA2Y2E0MzhmM2FlNDE1YmVkNDg2YjI4NWFiN2EwNWM5ODI5MzQ4Y2RhODA3OGQ1N2I5ODdjYWM2ODY4MDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bLo_gxVxDsvosbDcwjXki9FDAONhbCXLbe_nwCpYZKM)
> 
> ### Screenshot from Websurfx with the `strong` and `span` tags included
> ![2024-01-01_08-03](https://private-user-images.githubusercontent.com/31008330/293525765-ae2bb279-804f-4085-9ada-4970c217d973.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQwODI0MTYsIm5iZiI6MTcwNDA4MjExNiwicGF0aCI6Ii8zMTAwODMzMC8yOTM1MjU3NjUtYWUyYmIyNzktODA0Zi00MDg1LTlhZGEtNDk3MGMyMTdkOTczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAxVDA0MDgzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhMWRlYjcyZDViNTgwMzc5NTRkNWRmY2Y5YzNkYjcyZjg4MjkzYzg1OWQwM2M3YWY5N2E1ZjM4ZmRkNzRjZjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jXp2U_MU4Bq0Eaod09LZEyzoW8ZaJpZW0cCmoNB1bAQ)

Ok after seeing the picture I agree with you now that those `spans` are kind of useless and needs to be removed but for the `strong` tags I would suggest keeping them because I think as mentioned earlier when they get rendered on the search page they will add an aesthetic style like showing the user what in the description of the site matched their query which can help provide a better user experience. :slightly_smiling_face:

neon-mmd reviewed (changes requested):
Thanks :heart: for working on this PR. We really appreciate it :+1: but we would like to suggest a few changes before merge this PR. :slightly_smiling_face:

neon-mmd reviewed (approved):
Thanks :heart: again for working on this PR. We really appreciate it :+1: Since everything looks good. We will merge this PR right away :rocket: :slightly_smiling_face: .

neon-mmd commented:
@Mergifyio squash

neon-mmd committed:
✨ `Bing` for the search engine (#473)
fix: add closing curly brace

- accidentally removed from previous PR. i hope i did not ruin some
  build

* feat: implement mojeek engine

* Merge branch 'rolling' into FEAT_316_bing_search_engine

* ✨ feat: implement bing engine

* Fix: include cookie string to header_map

* Fix: include <strong> tags from upstream search

* Merge branch 'rolling' into FEAT_316_bing_search_engine
